### PR TITLE
feat(operator): postgresql database upgrade

### PR DESCRIPTION
**Pull request by zregvart**
We changed the version of the PostgreSQL database from 9.5 to 9.6. Any
time a a non patch version of the PostgreSQL database is increased
`pg_upgrade` must be run on the data directory (with old and new
binaries) to upgrade the data files.

Luckily, the Software Collections PostgreSQL image already contains
 script that checks for `POSTGRESQL_UPGRADE` environment variable and if
it's set to either `copy` or `hardlink` it will perform run the
`pg_upgrade` on startup. Caveat being that the `POSTGRESQL_UPGRADE` can
be only set if the upgrade is required; meaning that we can't set it and
leave it we have to manage it: set when needed and removed when not.

This adds a `DatabaseNeedsUpgrade` flag in the configuration and
depending on that flag sets the `POSTGRESQL_UPGRADE=copy` environment
variable. The `DatabaseNeedsUpgrade` flag is set if the container image
of the current `postgresql` container defined in the `DeploymentConfig`
is different from the image configured via
`Syndesis.Components.Database.Image` configuration.

On installation the state of the `syndesis-db` deployment is watched and
when it reaches the point where specified number of replicas equals the
ready number of replicas the `DeploymentConfig` is patched to remove the
`POSTGRESQL_UPGRADE=copy` environment variable.

This will cause a temporary outage, as changing the `DeploymentConfig`
results in the creation of new replication controller and termination of
the currently running `syndesis-db` pod.

Ref. https://issues.redhat.com/browse/ENTESB-12534

### CODE REVIEW

File: install/operator/pkg/cmd/internal/run/run.go [hunk 1]
```diff
@@ -124,7 +124,14 @@ func (o *options) run() error {
...
 
-	config, err := configuration.GetProperties(configuration.TemplateConfig, o.Context, nil, &v1beta1.Syndesis{})
+	cli, err := o.GetClient()
+	if err != nil {
+		return err
+	}
+
+	syndesis := &v1beta1.Syndesis{}
+	syndesis.SetNamespace(namespace)
+	config, err := configuration.GetProperties(configuration.TemplateConfig, o.Context, cli, syndesis)
```
[the remark this answers is not in the document]
Comment (zregvart) [reply install/operator/pkg/cmd/internal/run/run.go]: Without the namespace set [this bit](https://github.com/syndesisio/syndesis/pull/7953/files#diff-8f63b1158cd0c8188207de68b0013e69R286) fails.

**Comment by zregvart** [comment]
@phantomjinx can you push your changes to this fork/branch to incorporate them here.

I was also wondering if the `removePostgresUpgradeTrigger` is required, that is if the `DatabaseNeedsUpgrade` flag in the configuration would reset to `false` either way, it would not be needed as the template would render without `POSTGRESQL_UPGRADE=copy` and the operator would apply the change. Not sure about the life cycle of the configuration.

**Comment by phantomjinx** [comment]
> @phantomjinx can you push your changes to this fork/branch to incorporate them here.
> 
> I was also wondering if the `removePostgresUpgradeTrigger` is required, that is if the `DatabaseNeedsUpgrade` flag in the configuration would reset to `false` either way, it would not be needed as the template would render without `POSTGRESQL_UPGRADE=copy` and the operator would apply the change. Not sure about the life cycle of the configuration.

Yes it is needed. Unfortunately, the operator runs through the postUpgradeRun phase then the postUpgradeRunSucceed phase too quickly to allow for the database to be deployed and up before the DeploymentConfig is recreated, the existing db pod is terminated and a new one rolled out. As a result, the database is never migrated, resulting in a CrashLoopback.

**Comment by heiko-braun** [comment]
Very good work! Generalising the problem through `DatabaseNeedsUpgrade` is a good idea.

**Comment by zregvart** [comment]
I'll work on a backport to 1.9.x

**Comment by phantomjinx** [comment]
> I'll work on a backport to 1.9.x

Shouldn't be a problem as I cherry-picked to 1.9.x over the week while I was testing.

**Pull request merged**